### DISCUSSION TITLE
Re-attempt platform automerge

### DISCUSSION THREAD

**Discussion by straub** [renovatebot/renovate#26523] [Suggest an Idea]
### Tell us more.

Hey folks,

I've encountered a similar issue as https://github.com/renovatebot/renovate/discussions/25699, so I'd like to open the feature request mentioned there.

My use case is on GitHub, attempting to use their Merge Queue feature with Renovate. I have a dependency file that is very prone to merge conflicts and has many updates available (a Jenkins plugins.txt). I'm finding the GitHub Merge Queue actually works slightly _less_ smoothly than Renovate's native automerge, since when the Merge Queue ejects a PR that's developed merge conflicts during the process, it requires human intervention to make it back in, even after Renovate has rebased and resolved the conflict.

My understanding is that the root cause is that Renovate only attempts to add the PR to the merge queue when it's created, and not on future updates, via this code: https://github.com/renovatebot/renovate/blob/bde951257a79737351859d1c08db0ad214e61494/lib/modules/platform/github/index.ts#L1520

I also understand that infinite loops are possible in some configurations where checks are different in CI between the main branch and PRs, so having PRs re-added to the Merge Queue on updates by default might not be wise and could potentially waste CI resources forever. This is not a concern in my use case, as I have the same checks on the main branch as in PRs.

I don't think I exactly need Renovate to re-enable GitHub's automerge after _any_ push, but at least after Renovate's own pushes would be very helpful.

If you folks would be open to a new opt-in flag to enable another call to `tryPrAutomerge()` during `updatePr()`, I'd be happy to try my hand at implementing it!

Please let me know if you have any other questions, concerns, or guidance. And as always, thank you so much for all your hard work!

**Comment by rarkins** [comment]
Please go ahead! It would be handy to have such PRs added back to the queue as long as it doesn't result in any loops

**Reply by HonkingGoose** [reply]
Does Renovate have any way to see that's looping over the same PR? Here's what I'm guessing we'd need:

- Renovate opens PR
- User configures a (new) "stop trying to automerge after `x` attempts" config option, suppose the user sets this to `3`
- PR fails to get `platformAutomerged` with GitHub Merge Queue `3` times
- Renovate "sees" that it's tried 3 times, and thus exceeds it's alloted tries
- Renovate errors out on this PR with a suitable error message so the user can intervene

**Reply by straub** [reply]
In the [Issues Timeline API](https://docs.github.com/en/rest/issues/timeline?apiVersion=2022-11-28), I see an entry appear for `"event": "auto_merge_enabled"` when the app first creates the PR, then an `"event": "added_to_merge_queue"` entry attributed to the bot when checks pass and it's actually added to the queue. (There's also an `"event": "removed_from_merge_queue"` entry attributed to `github-merge-queue[bot]` when it eventually meets merge conflicts and is removed.)

So I think we could count instances of `"event": "added_to_merge_queue"` attributed to Renovate to get at the limit you're looking for, but it does require paging through a (potentially long) list.

**Reply by straub** [reply]
It's possible I'm also missing something here.. If Renovate were to re-add the PR to the merge queue only when it rebases (updates?) the PR (which already triggers a new build in CI by pushing commits), would that still be considered a loop that requires this limit placed on it? As far as I'm aware, there's no limit to the number of times Renovate will update a PR currently, so from a user's perspective it does seem slightly odd that we'd limit the attempts when GH Merge Queue is enabled.

**Reply by rarkins** [reply]
I think it's ok if we limit it to whenever Renovate rebases

**Reply by straub** [reply]
> limit it to whenever Renovate rebases

It looks to me like a simple way to accomplish this might be to implement a `refreshPr` method in the github platform and do a `tryPrAutomerge()` there. Does that sound reasonable to you? If so, I'll get to work on a PR!

**Reply by rarkins** [reply]
I'm not sure about the `refreshPr` naming specifically but probably it does need a new, optional function to be added to GitHub

**Reply by straub** [reply]
Sure, very open to suggestions! Trying to re-orient myself in Renovate's code, I went looking for how `updatePr()` might know a rebase just happened, and found `refreshPr()` instead. It looked like a good extension point in about the right place already: https://github.com/renovatebot/renovate/blob/4fc7554db9cda626f5e566140b082df10d1f2b81/lib/workers/repository/update/branch/index.ts#L574-L576

**Reply by straub** [reply]
I think I could rationalize the `refreshPr` naming as "we're refreshing the PR's GitHub automerge / merge queue status after rebasing", if that's convincing at all 😄

**Reply by straub** [reply]
I've submitted an initial PR for this: https://github.com/renovatebot/renovate/pull/26567

**Comment by straub** [comment]
I believe https://github.com/renovatebot/renovate/pull/26567 resolves this discussion, released in v37.255.0. Thanks!
